Refactor `EntityEvent` to support `ContainsEntity`

Zeenobit opened the pull request:
This is related to #21384

# Objective

The goal of this change is to enable `EntityEvent` to work with any Entity-like type which implements `ContainsEntity`.

## Solution

- Change return type of `EntityEvent::event_target` from `Entity` to `&impl ContainsEntity`
- Change all appropriate call sites to use `event_target().entity()` instead of `event_target()`
- Refactor `EntityEvent::event_target_mut` into `EntityEvent::set_event_target`

I'm not fully happy with this solution yet, but I'm opening the PR to discuss the options from here.

Mainly the issue resolved around `set_event_target`. To make this work, we need the underlying type to be constructible from an entity. This may not always be safe from user's perspective (it wouldn't be with `Instance<T>` for example).

Based on my understanding of the system, an event's target should is only mutated in the case of event propagation, which makes sense.

To solve this, I propose that we flag `EntityEvents` as **immutable** or not; or "not propagatable" or not (I'm open to other terminology! :P).

This would allow us to implement `set_event_target` as `unreachable()!` and throw an error instead if the user tries to propagate such an event. We could even maybe enforce this compile time but it'll require additional complexity.

## Testing

- Added `test_derive_entity_event` to cover all permutations

alice-i-cecile commented:
> &impl ContainsEntity

I'm reluctant to do this as it will break object safety. That will be a pretty widespread problem when trying to fix this though.

Zeenobit commented on examples/picking/sprite_picking.rs:
```diff
@@ -154,7 +154,7 @@ fn recolor_on<E: EntityEvent + Debug + Clone + Reflect>(
 ) -> impl Fn(On<E>, Query<&mut Sprite>) {
     move |ev, mut sprites| {
-        let Ok(mut sprite) = sprites.get_mut(ev.event_target()) else {
+        let Ok(mut sprite) = sprites.get_mut(ev.event_target().entity()) else {
```
Thinking more about it, yah, we could just isolate the change to `derive(EntityEvent)`. I'm thinking maybe we just call `ContainsEntity::entity(&self.#field)` in the `event_target` implementation.

alice-i-cecile commented:
> How do we feel about immutable entity events though?

Generally open to it, but haven't thought about this hard enough to be convinced that they're the only solution :)

Zeenobit commented:
I've updated the PR with the simplified implementation. Great idea! :)

> Generally open to it, but haven't thought about this hard enough to be convinced that they're the only solution :)

I'm open to suggestions!

From my usage of the engine so far, I have yet to find a use case for propagating events, which as I understand, is the only reason entity events are mutable.

So from my perspective, I could easily envision all entity events being immutable by default, and only mutable if `#[entity_event(propagate)]` is specified. Attempting to propagate an event not flagged with `#[entity_event(propagate)]` seems like a degenerate case to me, and so we could easily treat it as an error.

If we do that, we could then implement `set_event_target` with `unreachable!()` for immutable events, since it would never be called. Additionally, I would go as far as making `set_event_target` as unsafe. In my mind, this function should never be called externally.

Zeenobit commented:
I've added `SetEntityEventTarget` as discussed. Much cleaner implementation than what I had in mind! :)

For me, the only remaining issue is the `EntityCommand::trigger` call.

Should I split that into a separate PR or should I continue working in this PR?

Any opinions on how to refactor trigger?

In my opinion, the current implementation goes against Rust nomenclature. Typically in Rust we have functions like `verb` which take some data as input, or `verb_with` which take a constructor `Fn`.
We only have `trigger(f: impl FnOnce(Entity) -> E)`.

To fix this properly, I think `trigger` should just be `trigger(e: E)` and we add `trigger_with(f: impl FnOnce(Entity) -> E)`; but that would be a breaking change (which this PR already is? Maybe? 😅). I'm also open to other ideas on the naming.
